File: Chess.Game/Pieces/Pawn.cs
using System;
using System.Collections.Generic;

namespace Chess.Game.Pieces
{
    public class Pawn : IPiece
    {
        private int _multiplier => Color == Color.White ? 1 : -1;

        public int Moves { get; set; }
        public Color Color { get; set; }

        public IList<Location> GetLegalMoves(Location currentLocation)
        {
            throw new NotImplementedException();
        }

        public bool IsMoveLegal(Location current, Location newLocation)
        {
            if (newLocation.CurrentColor == Color)
                return false;

            var fileChange = Math.Abs((int)newLocation.File - (int)current.File);
            var rankChange = Math.Abs(newLocation.Rank - current.Rank);

            if (fileChange > 1 || rankChange > 2)
                return false;

            var legal = false;

            if (fileChange == 0)
            {
                var location = Color == Color.White ? current.Forward : current.Back;
                legal = location.Piece == null;

                if (rankChange == 2)
                {
                    location = Color == Color.White ? location.Forward : location.Back;
                    legal = Moves == 0 && location.Piece == null;
                }
            }
            else if (fileChange == 1)
            {
                if (rankChange == 1)
                {
                    if (Color == Color.White && newLocation.Rank > current.Rank ||
                        Color == Color.Black && newLocation.Rank < current.Rank)
                    {
                        legal = newLocation.CurrentColor != null && newLocation.CurrentColor != Color;
                    }
                }
                // TODO Add en passant
            }

            return legal;
        }
    }
}

File: Chess.Web/ClientApp/pieces/Pawn.ts
import { Location } from '../store/Game';
import Piece from './Piece';
import Color from './Color';

class Pawn implements Piece {
    private colorMultiplier: number;
    color: Color;
    moves: number;

    constructor(moves: number, color: Color) {
        this.moves = moves;
        this.color = color;
        this.colorMultiplier = color == Color.White ? 1 : -1;
    }

    isMoveLegal(current: Location, newLocation: Location, board: Location[][]) {
        if (newLocation.piece && newLocation.piece.color == this.color)
            return false;

        const fileChange = Math.abs(newLocation.file - current.file);
        const rankChange = this.colorMultiplier * (newLocation.rank - current.rank);

        if (fileChange > 1 || rankChange > 2 || rankChange < 0)
            return false;

        let legal = false;

        if (fileChange == 0)
        {
            var location = this.color == Color.White ? board[current.file - 1][current.rank] : board[current.file - 1][current.rank - 2];
            legal = !location.piece;

            if (rankChange == 2)
            {
                location = this.color == Color.White ? board[current.file - 1][current.rank] : board[current.file - 1][current.rank - 2];
                legal = !this.moves && !location.piece;
            }
        }
        else if (fileChange == 1)
        {
            if (rankChange == 1)
            {
                if (this.color == Color.White && newLocation.rank > current.rank ||
                    this.color == Color.Black && newLocation.rank < current.rank)
                {
                    legal = !!newLocation.piece && newLocation.piece.color != this.color;
                }
            }
        }

        return legal;
    };

    getImageName() {
        const image = this.color == Color.White ? "white_pawn" : "black_pawn";

        return image;
    }
}

export default Pawn;
